x-pack/plugins/security_solution/server/lib/detection_engine/rules/bulk_action_edit.ts

```diff
@@ -40,24 +42,38 @@ export const applyBulkActionEditToRule = (
       break;
 
     // index_patterns actions
-    // index is not present in all rule types(machine learning). But it's mandatory for the rest.
-    // So we check if index is present and only in that case apply action
+    // index pattern is not present in machine learning rule type, so we throw error on it
     case BulkActionEditType.add_index_patterns:
-      if (rule.params && 'index' in rule.params) {
-        rule.params.index = addItemsToArray(rule.params.index ?? [], action.value);
-      }
+      invariant(
+        rule.params.type !== 'machine_learning',
+        "Index patterns can't be added. Machine learning rule doesn't have index patterns property"
+      );
+
+      rule.params.index = addItemsToArray(rule.params.index ?? [], action.value);
       break;
 
     case BulkActionEditType.delete_index_patterns:
-      if (rule.params && 'index' in rule.params) {
-        rule.params.index = deleteItemsFromArray(rule.params.index ?? [], action.value);
-      }
+      invariant(
+        rule.params.type !== 'machine_learning',
+        "Index patterns can't be deleted. Machine learning rule doesn't have index patterns property"
+      );
+
+      rule.params.index = deleteItemsFromArray(rule.params.index ?? [], action.value);
+
+      invariant(
+        rule.params.index.length !== 0,
+        "Can't delete all index patterns. At least one index pattern must be left"
+      );
       break;
 
     case BulkActionEditType.set_index_patterns:
-      if (rule.params && 'index' in rule.params) {
-        rule.params.index = action.value;
-      }
+      invariant(
+        rule.params.type !== 'machine_learning',
+        "Index patterns can't be overwritten. Machine learning rule doesn't have index patterns property"
+      );
+      invariant(action.value.length !== 0, "Index patterns can't be overwritten with empty list");
```

Comment:
Thank you for adding all these checks as invariants, looks great!
Curious, however, isn't `action.value.length !== 0` already guaranteed by the request schema? (same for `add_index_patterns` and `delete_index_patterns`)

Reply:
no, it's not

It uses https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts#L105

I guess, we might try to go ahead and use NonEmptyArray.
I will definitely look into in in next iterations of bulk edit :+1